crates/sui-core/src/checkpoints/checkpoint_executor/mod.rs

```diff
@@ -442,6 +526,56 @@ async fn execute_transactions(
         })
         .collect();
 
+    // Pull out change_epoch tx (if exists) so that we can special case its execution last
+    let end_of_epoch = if let Some(index) = executable_txns
+        .iter()
+        .position(|tx| tx.transaction_data().is_change_epoch_tx())
+    {
+        executable_txns.remove(index);
+        all_tx_digests.remove(index);
+        execution_digests.remove(index);
+        true
+    } else {
+        false
+    };
+
+    let effects = execute_transactions_impl(
+        execution_digests,
+        all_tx_digests.clone(),
+        executable_txns,
+        authority_store.clone(),
+        epoch_store.clone(),
+        transaction_manager,
+        log_timeout_sec,
+        checkpoint,
+    )
+    .await?;
+
+    // if end of epoch checkpoint, we must finalize the checkpoint after executing
+    // the change epoch tx, which is done after all other checkpoint execution
+    if !end_of_epoch {
+        finalize_checkpoint(
+            authority_store,
+            all_tx_digests,
+            epoch_store,
+            checkpoint_sequence,
+            accumulator,
+            effects,
+        )?;
+    }
```

Comment:
These seem to be better fit in the parent function execute_checkpoint, given that they are all dealing with checkpoint specific logic.

Reply:
Fixed.